src/plugins/interactive_setup/server/routes/enroll.ts

```diff
@@ -6,26 +6,143 @@
+            apiKey: schema.string(),
+            caFingerprint: schema.string(),
+          }),
+          schema.object({
+            hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+            username: schema.string(),
+            password: schema.string(),
+            caCert: schema.string(),
+          }),
+        ]),
       },
       options: { authRequired: false },
     },
     async (context, request, response) => {
-      return response.forbidden({
-        body: { message: `API is not implemented yet.` },
-      });
+      const client = core.elasticsearch
+        .createClient('enroll', {
+          hosts: request.body.hosts,
+          ssl: { verificationMode: 'none' },
+          // caFingerprint: request.body.caFingerprint,
+        })
+        .asScoped({
+          headers: {
+            authorization:
+              'apiKey' in request.body
+                ? `ApiKey ${btoa(request.body.apiKey)}`
+                : `Basic ${btoa(`${request.body.username}:${request.body.password}`)}`,
+          },
+        });
+
+      const configPath = initializerContext.env.mode.dev
+        ? initializerContext.env.configs.find((fpath) => path.basename(fpath).includes('dev'))
+        : initializerContext.env.configs[0];
+
+      if (!configPath) {
+        logger.error('Cannot find config file');
+        return response.customError({ statusCode: 500, body: 'Cannot find config file.' });
+      }
+      // TODO: Use fingerprint/timestamp as file name
+      const caPath = path.join(path.dirname(configPath), 'ca.crt');
+
+      try {
+        await Promise.all([
+          fs.access(configPath, constants.W_OK),
+          fs.access(path.dirname(caPath), constants.W_OK),
+        ]);
+
+        const { body } = await client.asCurrentUser.transport.request({
+          method: 'GET',
+          path: '/_security/enroll/kibana',
+        });
+
+        const certificateAuthority = generateCertificate(body.http_ca);
+
+        // Ensure we can connect using CA before writing config
+        const verifyConnectionClient = core.elasticsearch.createClient('verifyConnection', {
+          hosts: request.body.hosts,
+          username: 'kibana_system',
+          password: body.password,
+          ssl: { certificateAuthorities: [certificateAuthority] },
+        });
+
+        await verifyConnectionClient.asInternalUser.security.authenticate();
+
+        await Promise.all([
+          fs.appendFile(
+            configPath,
+            generateConfig(request.body.hosts, 'kibana_system', body.password, caPath)
+          ),
+          fs.writeFile(caPath, certificateAuthority),
+        ]);
+
+        completeSetup({ shouldReloadConfig: true });
+
+        return response.noContent();
+      } catch (error) {
+        logger.error(error);
+        return response.customError({ statusCode: 500 });
+      }
     }
   );
 }
+
+export function generateCertificate(pem: string) {
```

Comment:
__question:__ is this some standard approach? Do we have a reliable existing npm package that can help us with this or ...?

Reply:
There's `node-forge` which can handle most certificate format conversions but it felt a bit heavy for this single use case since its a full openssl implementation. 

Node v16 has a [X509Certificate](https://nodejs.org/api/crypto.html#crypto_class_x509certificate) class that might be able to format base64 to PEM but we're on v14 LTS.

Reply:
> There's node-forge which can handle most certificate format conversions but it felt a bit heavy for this single use case since its a full openssl implementation.

Agree, it's too much.

> Node v16 has a X509Certificate class that might be able to format base64 to PEM but we're on v14 LTS.

That's great! Would you mind leaving a comment to try to upgrade our code to use `X509Certificate` once it's available?